src/pmtiles_adapter.js

```diff
@@ -0,0 +1,145 @@
+    const buffer = Buffer.alloc(length);
+    await ReadFileBytes(this.fd, buffer, offset);
+    return { data: BufferToArrayBuffer(buffer) };
+  }
+};
+
+const ReadFileBytes = async (fd, buffer, offset) => {
+  return new Promise((resolve, reject) => {
+    fs.read(fd, buffer, 0, buffer.length, offset, (err) => {
+      if (err) {
+        return reject(err);
+      }
+      resolve();
+    });
+  });
+};
+
+export const PMtilesOpen = (FilePath) => {
+  let pmtiles = undefined;
+  let fd = undefined;
+
+  if (isValidHttpUrl(FilePath)) {
+    const source = new PMTiles.FetchSource(FilePath);
+    pmtiles = new PMTiles.PMTiles(source);
+  } else {
+    fd = fs.openSync(FilePath, 'r');
+    const source = new PMTilesFileSource(fd);
+    pmtiles = new PMTiles.PMTiles(source);
+  }
+  return { pmtiles: pmtiles, fd: fd };
+};
+
+export const PMtilesClose = (fd) => {
+  fs.closeSync(fd);
+};
+
+export const GetPMtilesInfo = async (pmtiles) => {
+  const header = await pmtiles.getHeader();
+  const metadata = await pmtiles.getMetadata();
+
+  //Add missing metadata from header
+  metadata['format'] = GetPmtilesTileType(header.tileType).type;
+  metadata['minzoom'] = header.minZoom;
+  metadata['maxzoom'] = header.maxZoom;
+
+  if (header.minLon && header.minLat && header.maxLon && header.maxLat) {
+    metadata['bounds'] = [
+      header.minLon,
+      header.minLat,
+      header.maxLon,
+      header.maxLat,
+    ];
+  } else {
+    metadata['bounds'] = [-180, -85.05112877980659, 180, 85.0511287798066];
+  }
+
+  if (header.centerZoom) {
+    metadata['center'] = [
+      header.centerLon,
+      header.centerLat,
+      header.centerZoom,
+    ];
+  } else {
+    metadata['center'] = [
+      header.centerLon,
+      header.centerLat,
+      parseInt(metadata['maxzoom']) / 2,
+    ];
+  }
+
+  return metadata;
+};
+
+export const GetPMtilesTile = async (pmtiles, z, x, y) => {
+  const header = await pmtiles.getHeader();
+  const TileType = GetPmtilesTileType(header.tileType);
+  let zxyTile = await pmtiles.getZxy(z, x, y);
+  if (zxyTile && zxyTile.data) {
+    zxyTile = ArrayBufferToBuffer(zxyTile.data);
+  } else {
+    zxyTile = undefined;
+  }
+  return { data: zxyTile, header: TileType.header };
+};
+
+const GetPmtilesTileType = (typenum) => {
+  let head = {};
+  let tileType;
+  switch (typenum) {
+    case 0:
+      tileType = 'Unknown';
+      break;
+    case 1:
+      tileType = 'pbf';
+      head['Content-Type'] = 'application/x-protobuf';
+      break;
+    case 2:
+      tileType = 'png';
+      head['Content-Type'] = 'image/png';
+      break;
+    case 3:
+      tileType = 'jpeg';
+      head['Content-Type'] = 'image/jpeg';
+      break;
+    case 4:
+      tileType = 'webp';
+      head['Content-Type'] = 'image/webp';
+      break;
+    case 5:
+      tileType = 'avif';
+      head['Content-Type'] = 'image/avif';
+      break;
+  }
+  return { type: tileType, header: head };
+};
+
+const BufferToArrayBuffer = (buffer) => {
```

Comment:
the pmtiles code seems to want ArrayBuffers . And the old rendering code works with buffers.

Initially I had gone with this code for the source, which used slice like you said
https://github.com/maptiler/tileserver-gl/blob/ccd1dbf343c28cc899a5e0eb8f323c7f5f9b4b7f/src/main.js#L78-L89
but I basically had to pass it the full file as an ArrayBuffer for it to work. I couldn't find a good way to read a file to array buffer in node that didn't make the loading take forever. Anything I tried to do to load the file went to a buffer first, which had a 2GB limit. That is why I switched to the file open based approach, which was much faster because it didn't need to load the 

The converting back sounds like it be improved based on what you said, i'll add that.

Reply:
I was thinking of this:

```js
function bufferToArrayBuffer(buffer) {
  return buffer.buffer.slice(buffer.byteOffset, buffer.byteOffset + buffer.byteLength);
}

function arrayBufferToBuffer(arrayBuffer) {
  return Buffer.from(arrayBuffer);
}
```

Reply:
Makes sense. I didn't think of using slice with the smaller buffer from read file.

Reply:
I've updated this to use the inbuild buffer functions